swift/extractor/invocation/CodeQLDiagnosticsConsumer.cpp

```diff
@@ -0,0 +1,97 @@
+#include "swift/extractor/invocation/CodeQLDiagnosticsConsumer.h"
+#include "swift/extractor/trap/TrapDomain.h"
+#include "swift/extractor/trap/generated/TrapEntries.h"
+
+#include <swift/AST/DiagnosticEngine.h>
+#include <swift/Basic/SourceManager.h>
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/raw_ostream.h>
+#include <string>
+
+using namespace codeql;
+
+static int diagnosticsKind(const swift::DiagnosticInfo& diagInfo) {
+  switch (diagInfo.Kind) {
+    case swift::DiagnosticKind::Error:
+      return 1;
+    case swift::DiagnosticKind::Warning:
+      return 2;
+
+    case swift::DiagnosticKind::Note:
+      return 3;
+
+    case swift::DiagnosticKind::Remark:
+      return 4;
+  }
+  return 0;
+}
+
+static std::filesystem::path getFilePath(std::string_view path) {
+  // TODO: this needs more testing
+  // TODO: check canonicalization of names on a case insensitive filesystems
+  // TODO: make symlink resolution conditional on CODEQL_PRESERVE_SYMLINKS=true
+  std::error_code ec;
+  auto ret = std::filesystem::canonical(path, ec);
+  if (ec) {
+    std::cerr << "Cannot get real path: " << std::quoted(path) << ": " << ec.message() << "\n";
+    return {};
+  }
+  return ret;
+}
+
+static void attachLocation(TrapDomain& trap,
+                           swift::SourceManager& sourceManager,
+                           const swift::DiagnosticInfo& diagInfo,
+                           DiagnosticsTrap& locatable) {
+  auto loc = diagInfo.Loc;
+  if (!loc.isValid()) {
+    return;
+  }
+  auto filepath = getFilePath(sourceManager.getDisplayNameForLoc(loc));
+  FilesTrap file;
+  file.id = trap.createLabel<FileTag>();
+  file.name = filepath;
+  trap.emit(file);
+
+  LocationsTrap location;
+  location.id = trap.createLabel<LocationTag>();
+  location.file = file.id;
+  std::tie(location.start_line, location.start_column) =
+      sourceManager.getLineAndColumnInBuffer(loc);
+  std::tie(location.end_line, location.end_column) = sourceManager.getLineAndColumnInBuffer(loc);
+  trap.emit(location);
+  trap.emit(LocatableLocationsTrap{locatable.id, location.id});
```

Comment:
you are emitting non-final classes, that's your issue. `File` is now a super class of `DbFile` (should really be named `RawFile` or something) and synthesized `UnknownFile`. Same goes for `Location`. Also, you are not deduplicating with trap keys. I guess you took this code from a previous version of the extractor, prior to the synthesized `UnknownFile` and `UnknownLocation`.
```suggestion
  DbFile file{trap.createLabel<FileTag>(filepath)};
  file.name = filepath;
  trap.emit(file);
  DbLocation loc{
  file.name = filepath;
  trap.emit(file);

  DbLocation entry{{}};
  entry.file = fetchLabel(file);
  std::tie(entry.start_line, entry.start_column) = sourceManager.getLineAndColumnInBuffer(start);
  std::tie(entry.end_line, entry.end_column) = sourceManager.getLineAndColumnInBuffer(end);
  entry.id = trap.createLabel<DbLocationTag>('{', entry.file, "}:", entry.start_line, ':',
                                             entry.start_column, ':', entry.end_line, ':',
                                             entry.end_column);
  emit(entry);
  emit(LocatableLocationsTrap{locatableLabel, entry.id});
```
We should probably find a way to share this with `SwiftDispatcher`. Maybe a separate `SwiftLocationExtractor` for locations and files, with its own cache?

Reply:
the problem is that I think the database will be flooded with duplicate files (and possibly locations, but that's less problematic). It's not a question of performance, but of database consistency :slightly_smiling_face:

Reply:
in any case, we definitely have to share the code for emitting files and locations, so once that is done proper trap keys will be emitted